examples/TCP/IBerkeleySocket.cs

```diff
@@ -0,0 +1,162 @@
+/*
+Berkeley sockets interface.
+Nik Sultana, Cambridge University Computer Lab, December 2016
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+*/
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using System.Net;
+using Pax;
+
+namespace Pax_TCP {
+  /* FIXME right now I only care about servers, therefore don't include
+     functions like "connect", "gethostbyname", and "gethostbyaddr" in the API.
+     For simplicity, I don't include "getsockopt", "setsockopt", "select", and
+     "poll" either.
+     Also i exclude "sendto", "recvfrom", "send", and "recv", again for prototyping.
+  */
+  public interface IBerkeleySocket {
+    Result<SockID> socket (Internet_Domain domain, Internet_Type type, Internet_Protocol prot);
+    Result<Unit> bind (SockID sid, SockAddr_In address);
+    Result<Unit> listen (SockID sid); // NOTE we let the "backlog" parameter be implicit for prototyping reasons; it's a parameter of TCP, not the interface.
+    Result<SockID> accept (SockID sid, out SockAddr_In address);
+    Result<int> write (SockID sid, byte[] buf, uint count);
+    Result<int> read (SockID sid, byte[] buf, uint count);
+    Result<Unit> close (SockID sid);
+  }
+
+  public interface IActiveBerkeleySocket: IBerkeleySocket, IActive {}
+
+  public enum Unit { Value };
+
+  // FIXME many constant values are excluded, e.g., various error codes.
+  //       i only included the ones that i think i'll use in the prototype.
+  public enum Internet_Domain {AF_Inet};
+  public enum Internet_Type {Sock_Stream};
+  public enum Internet_Protocol {TCP};
+  public enum Error {EACCES, EADDRINUSE, EBADF, EINVAL, ENOTSOCK, EOPNOTSUPP, EINTR, EIO,
+    EAGAIN, EWOULDBLOCK, EDESTADDRREQ, EDQUOT, EFAULT, EFBIG, ENOSP, EPERM, EPIPE};
+
+  // FIXME this is crude -- since it's possible to have "stale" references to
+  //       "old" sockets (i.e., offsets in the TCB array) get pointing to "new"
+  //       sockets (when the TCB at that offset happens to be in the expected
+  //       state). Perhaps this could be fixed by adding a unique number
+  //       to each SockID instance, to ensure freshness.
+  public class SockID {
+    public SockID (uint sockid) {
+      this.sockid = sockid;
+    }
+
+    public uint sockid {
+      get;
+      set;
+    }
+  }
+
+  public class Result<T> {
+    public readonly T result;
+    public readonly bool erroneous = false;
+    public readonly Error? error;
+
+    public Result (T result, Error? error) {
+      this.result = result;
+      this.error = error;
+      if (error != null) {
+        erroneous = true;
+      }
+    }
+
+    public T value_exc () {
+      if (erroneous) {
+        throw new Exception(this.ToString());
+      }
+
+      return result;
+    }
+
+    override public string ToString() {
+      string result;
+
+      if (error == null) {
+        return this.result.ToString();
+      }
+
+      switch (error) {
```

Comment:
Could this be replaced with ```error.ToString()```? It would return the numerical value when not a defined enum value, but if that matters perhaps the following extension method is concise enough?

``` csharp
/* In some aptly named static class */
/// <summary>
/// Converts the value of this enum instance to its string value (name) if possible, or return a default otherwise.
/// </summary>
/// <param name="e">The enum value.</param>
/// <param name="defaultStr">The string to return if the value is undefined in the enum.</param>
/// <remarks>See <seealso cref="Enum.ToString"/> for details. </remarks>
public static string AsString(this Enum e, string defaultStr)
{
  string eStr = e.ToString("G"); // NOTE will return csv if the enum has the Flags attribute
  string eNum = e.ToString("D"); // The numerical value
  if (eStr != eNum) 
    return eStr;
  else
    return defaultStr;
}
...
return error.AsString("(Unknown error)");
```